modules/core/src/main/java/org/apache/ignite/internal/causality/AbstractVersionedValue.java

```diff
@@ -0,0 +1,359 @@
+        }
+
+        List<CompletableFuture<T>> futuresToComplete = List.of();
+
+        CompletableFuture<T> previousCompleteFuture = null;
+
+        for (CompletableFuture<T> future : headMap.descendingMap().values()) {
+            if (future.isDone()) {
+                previousCompleteFuture = future;
+
+                break;
+            }
+
+            // Lazy initialization.
+            if (futuresToComplete.isEmpty()) {
+                futuresToComplete = new ArrayList<>();
+            }
+
+            futuresToComplete.add(future);
+        }
+
+        // We found the first complete future, but there are no incomplete futures up to the actual token, so there's nothing to do.
+        if (futuresToComplete.isEmpty()) {
+            return previousCompleteFuture;
+        }
+
+        if (previousCompleteFuture == null) {
+            // "complete" method has never been called before, use the default value as the original.
+            T defaultValue = getDefault();
+
+            futuresToComplete.forEach(f -> f.complete(defaultValue));
+        } else {
+            assert previousCompleteFuture.isDone();
+
+            // Create an effectively final variable.
+            List<CompletableFuture<T>> futuresToCompleteCopy = futuresToComplete;
+
+            previousCompleteFuture.whenComplete((v, t) -> {
+                if (t != null) {
+                    futuresToCompleteCopy.forEach(f -> f.completeExceptionally(t));
+                } else {
+                    futuresToCompleteCopy.forEach(f -> f.complete(v));
+                }
+            });
+        }
+
+        return previousCompleteFuture;
+    }
+
+    /**
+     * Updates the current token. Must be called under the {@link #readWriteLock writeLock}.
+     */
+    private void setActualToken(long causalityToken) {
+        assert actualToken < causalityToken
+                : format("Token must be greater than actual [token={}, actual={}]", causalityToken, actualToken);
+
+        actualToken = causalityToken;
+    }
+
+    /**
+     * Trims the storage to history size. Must be called under the {@link #readWriteLock writeLock}.
+     */
+    private void trimHistory() {
+        NavigableMap<Long, CompletableFuture<T>> oldTokensMap = history.headMap(actualToken, true);
+
+        if (oldTokensMap.size() <= maxHistorySize) {
+            return;
+        }
+
+        Iterator<Map.Entry<Long, CompletableFuture<T>>> it = oldTokensMap.entrySet().iterator();
+
+        for (int i = oldTokensMap.size(); i > maxHistorySize; i--) {
+            Map.Entry<Long, CompletableFuture<T>> next = it.next();
+
+            // All futures must be explicitly completed before history trimming occurs.
+            assert next.getValue().isDone();
+
+            it.remove();
+        }
+    }
+
+    /**
+     * Copies the state of the {@code from} future to the incomplete {@code to} future.
+     */
+    private void copyState(CompletableFuture<T> from, CompletableFuture<T> to) {
```

Comment:
The implementation of this method is trivial and short, I don't think having yet another utility method will be beneficial. Or at least it will bloat this huge PR even more

Reply:
Ok